pkg/controller/persistentvolume/persistentvolume_controller.go

```diff
@@ -0,0 +1,1438 @@
+					// User asked for a specific PV, retry later
+					if _, err = ctrl.updateClaimPhase(claim, api.ClaimPending); err != nil {
+						return err
+					}
+					return nil
+				} else {
+					// This should never happen because someone had to remove
+					// annWasEverBound annotation on the claim.
+					glog.V(5).Infof("Synchronizing unbound PersistentVolumeClaim[%s]: volume already bound to different claim %q by controller, THIS SHOULD NEVER HAPPEN", claimToClaimKey(claim), claimrefToClaimKey(volume.Spec.ClaimRef))
+					return fmt.Errorf("Invalid binding of claim %q to volume %q: volume already claimed by %q", claimToClaimKey(claim), claim.Spec.VolumeName, claimrefToClaimKey(volume.Spec.ClaimRef))
+				}
+			}
+		}
+	}
+}
+
+func (ctrl *PersistentVolumeController) syncBoundClaim(claim *api.PersistentVolumeClaim) error {
+	// hasAnnotation(pvc, annWasEverBound)
+	// This PVC has previously been bound
+	// OBSERVATION: pvc is not "Pending"
+	// [Unit test set 3]
+	if claim.Spec.VolumeName == "" {
+		// Claim was bound before but not any more.
+		if _, err := ctrl.updateClaimPhaseWithEvent(claim, api.ClaimLost, api.EventTypeWarning, "ClaimLost", "Bound claim has lost reference to PersistentVolume. Data on the volume is lost!"); err != nil {
+			return err
+		}
+		return nil
+	}
+	obj, found, err := ctrl.volumes.store.GetByKey(claim.Spec.VolumeName)
+	if err != nil {
+		return err
+	}
+	if !found {
+		// Claim is bound to a non-existing volume.
+		if _, err = ctrl.updateClaimPhaseWithEvent(claim, api.ClaimLost, api.EventTypeWarning, "ClaimLost", "Bound claim has lost its PersistentVolume. Data on the volume is lost!"); err != nil {
+			return err
+		}
+		return nil
+	} else {
+		volume, ok := obj.(*api.PersistentVolume)
+		if !ok {
+			return fmt.Errorf("Cannot convert object from volume cache to volume %q!?", claim.Spec.VolumeName)
+		}
+
+		glog.V(5).Infof("Synchronizing bound PersistentVolumeClaim[%s]: volume %q found: %s", claimToClaimKey(claim), claim.Spec.VolumeName, ctrl.getVolumeStatusForLogging(volume))
+		if volume.Spec.ClaimRef == nil {
+			// Claim is bound but volume has come unbound.
+			// Or, a claim was bound and the controller has not received updated
+			// volume yet. We can't distinguish these cases.
+			// Bind the volume again and set all states to Bound.
+			glog.V(5).Infof("Synchronizing bound PersistentVolumeClaim[%s]: volume is unbound, fixing")
+			if err = ctrl.bind(volume, claim); err != nil {
+				// Objects not saved, next syncPV or syncClaim will try again
+				return err
+			}
+			return nil
+		} else if volume.Spec.ClaimRef.UID == claim.UID {
+			// All is well
+			// NOTE: syncPV can handle this so it can be left out.
+			// NOTE: bind() call here will do nothing in most cases as
+			// everything should be already set.
+			glog.V(5).Infof("Synchronizing bound PersistentVolumeClaim[%s]: Everything is OK", claimToClaimKey(claim))
```

Comment:
I feel like there is something fundamentally unlucky about saying 'Everything is OK' in a log message :game_die: :game_die: